### PR TITLE
Marshaler registry

### PR DESCRIPTION
**Pull request by obeattie**
Add a marshaler registry, which is used to determine which marshaler or unmarshaler to use for a given request or response. This was previously hard-coded into the client/server, but I want higher-level code to be able to supply their own, while still benefiting from automatic (un)marshaling.

### CODE REVIEW

File: server/endpoint.go [hunk 1]
```diff
@@ -23,16 +24,20 @@ type Endpoint struct {
 	Response proto.Message
 }
 
+func (e Endpoint) unmarshaler(req mercury.Request) tmsg.Unmarshaler {
+	return marshaling.Unmarshaler(req.Headers()[marshaling.ContentTypeHeader], e.Request)
+}
+
 // Handle takes an inbound Request, unmarshals it as a protobuf, dispatches it to the handler, and serialises the
```
Comment (mattheath) [inline server/endpoint.go]: _"unmarshals it as a protobuf"_

File: request.go [hunk 2]
```diff
@@ -41,16 +38,16 @@ func responseFromRequest(req Request, body proto.Message) Response {
...
-		default:
-			err = tmsg.ProtoMarshaler().MarshalBody(rsp)
+
+		ct := req.Headers()[marshaling.AcceptHeader]
+		marshaler := marshaling.Marshaler(ct)
+		if marshaler == nil { // Fall back to proto
+			marshaler = marshaling.Marshaler(marshaling.ProtoContentType)
 		}
-		if err != nil {
+		if marshaler == nil {
```
Comment (mattheath) [inline request.go]: Can this fail a second time as you fall back to proto just above? The proto marshaler is initialised at a package level in the `marshalerRegistry`, and there isn't a method to remove it.

File: client/client.go [hunk 3]
```diff
@@ -146,7 +147,10 @@ func (c *client) performCall(call clientCall, middleware []ClientMiddleware, tra
 		// call's response nil
 		if rsp.IsError() {
 			errRsp := rsp.Copy()
-			unmarshaler := tmsg.ProtoUnmarshaler(&tperrors.Error{})
+			unmarshaler := marshaling.Unmarshaler(rsp.Headers()[marshaling.ContentTypeHeader], &tperrors.Error{})
+			if unmarshaler == nil {
+				unmarshaler = marshaling.Unmarshaler(marshaling.ContentTypeHeader, &tperrors.Error{})
```
Comment (mattheath) [inline client/client.go]: Not sure I understand this - `marshaling.ContentTypeHeader` is the string `"Content-Type"`?

File: server/endpoint.go (hunk 1 above)
Comment (mattheath) [inline server/endpoint.go]: sorry, clearly looking at an out of date diff :unamused:

File: client/client.go (hunk 3 above)
Comment (mattheath) [inline client/client.go]: Again, must have looked at an older commit. Sorry!

File: server/endpoint.go [hunk 4]
```diff
@@ -1,40 +1,53 @@
...
+				return
+			}
 		}
 	}
 
-	return e.Handler(req)
+	defer func() {
+		if v := recover(); v != nil {
+			log.Criticalf("[Mercury:Server] Recovered from handler panic for request %s: %v", req.Id(), v)
+			rsp, err = nil, terrors.InternalService("Unhandled exception in endpoint handler")
```
Comment (jonashuckestein) [inline server/endpoint.go]: Would love to attach the panic information to the error somehow. May be worth having terrors.Panic in the future

**Comment by jonashuckestein** [comment]
LGTM. One tiny thing that I'm happy to add in a different PR

**Pull request merged**